React optimization + event based processing for tanks, canisters and pipelines

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

I've converted tanks, canisters, and pipelines fully over to reaction based processing. That means they won't process at roundstart, and if they stop being "interesting", they'll stop. Part of this is making update_parents() only get called when gas is actually moved, rather then all the time.

The idea is to force anything that interacts with a tank/canister/pipeline's gasmixture to use return_air rather then directly accessing it. I need to do a second pass on anything that might be caching it, but I don't think many if any things do that.

I've also included react_hybrid in this pr, which unlike react() gets more expensive with gas "in the mix" count rather then total reaction count. It's cheaper then the old react proc for low gas type mixes, is about parity for 8 gas types, and not too much worse for 21, our max. I've hardcoded the nobilium check too, because doing it in a modular way was slower. If anyone tries to make nobilium 2.0 they can suck it.

Basically, instead of giving each reaction its own priority, we split them into 4 priority groups, pre formation, formation, post formation, and fire. they should just about cover everything you'd want to do with a reaction. This also means you can't return `STOP_REACTIONS` to stop all reactions normally, but I'm fine with that. 

Outside of that I've done some simple optimizations to both react() procs, including as anythings and the removal of an unused check for energy.

Have some profiles from a local:
"Best" case scenario, so 2 gas types, o2 and n2, and no actual reactions being tested or performed.
![image](https://user-images.githubusercontent.com/58055496/114682841-534cd780-9cc4-11eb-908e-b256838dcc53.png)

"Middling" case, so a plasma + co2 + n2o + o2 fire started by opening one plasma and one o2 canister in meta toxins, waiting for them to settle, then igniting them with the apc's overload button. Profiling continued until the floor melted away and all gas dissipated.
![image](https://user-images.githubusercontent.com/58055496/114683175-a888e900-9cc4-11eb-9462-daa651e0b506.png)

"Worst" case, so all 21 gases in one mixture. (This was without the nobilium hardcode, but it should be close to trivial)
![image](https://user-images.githubusercontent.com/58055496/114683277-bf2f4000-9cc4-11eb-95ba-b8510da73c29.png)

These profiles were done on a local with tank and canister reactions disabled after running the following sqdl commands, and waiting for things to equalize. Oh and react_kills_prioirty is what it says on the tin, it just ignored all forms of priority outside of a nobilium check, was included in those tests as a best case senario
```
DELETE /mob/living ///Mobs can open doors, and monkeys breath the bastards
DELETE /obj/machinery/atmospherics ///For obvious reasons
```

I'm looking to test both systems on master with all its havoc and gas diffs, I've already tested the reaction change on master, but there's too many useless reacts to get much good data from that, so let's try em both at once and see what blows up yeah?

There's a potential hole I'm opening up here by the way, if people figure out how to force reactions to only fire when they want them to rather then all the time, we'll have a new sort of bomb. If that happens it should be reasonably easy to resolve, but I figured I'd let you know.

### Current Benefits
Makes the best case react() call ~3x less expensive, while unshackling us from a limited reaction count.
Brings the default load from pipeline processing down from 24 to 16, on meta.
Make react() calls more sane, reduces the amount of unneeded pipeline equalization.
Sets me up for more work.
 
### Some further improvements

We could in theory remove the fire reaction group, but that makes things a bit fiddly
Someone (Me) Should really convert vents and scrubbers to atmos_sensitive, to remove all the unneeded processes they produce. seriously they're a good portion of the 50ms default load atmos has at roundstard

### Post testing
This was close to a micro op, but it's very helpful even outside of that because it removes the weight of caring about reaction length in the absolute sense. Still fucking faster tho.
![image](https://user-images.githubusercontent.com/58055496/116349635-4987a180-a7a5-11eb-9806-8c8e294e9bc0.png)

There's a slight cost to this, in that people need to respect the rules around working with tank/canister gas mixtures, I've added a comment to this effect and written a little blurb at the end of Atmospherics.md on the topic

## Why It's Good For The Game

**NO UNNEEDED MOVEMENTS**
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: LemonInTheDark, duncathan 
refactor: Gas reactions should be cheaper in the standard case
refactor: Moves canisters, pipelines, and tanks to event based processing, if you see something fail to "update" when it really should let me know yeah?
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Makes tanks only process when needed, rather then doing it constantly

LemonInTheDark committed:
Makes portable atmos machines only process when needed, makes adding and removing atmos machinery cheaper, makes the tank processing logic actually work properly

LemonInTheDark committed:
Makes pipelines only react() when their mix changes, fixes a ton of misuses of update_parents that were causing about 10ms of load for atmos just from reconcile_air()

LemonInTheDark committed:
Adds in a new reaction framework alongside the old one for profiling related reasons

LemonInTheDark committed:
whoops

Ghilker commented on code/game/objects/items/flamethrower.dm:
```diff
@@ -215,12 +215,14 @@
+
 	if(air_transfer.gases[/datum/gas/plasma])
-		air_transfer.gases[/datum/gas/plasma][MOLES] *= 5
+		air_transfer.gases[/datum/gas/plasma][MOLES] *= 5 //Suffering
```
we should just remove that *= 5

LemonInTheDark replied:
it's good for gameplay thoooo

duncathan reviewed (approved):
I'm duncathan salt, and I approve this message

LemonInTheDark commented:
![image](https://user-images.githubusercontent.com/58055496/116333848-159e8300-a789-11eb-8049-9931c389e95e.png)
it's a micro op, but it also means we don't scale by reaction count anymore, which is quite nice

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into lemon-gets-mad-at-react

LemonInTheDark committed:
Cleanup, removes the react profiling code, I've chosen a direction to go here

LemonInTheDark committed:
Cleans up some code, adds comments describing how to interact with portable atmos machines/tanks and their
mixes, adds a blurb to Atmospherics.md about the topic as a whole, leaves a line of bread for someone to follow
if I get hit by a bus in 2 days (Knock on wood)

LemonInTheDark committed:
Removes priority from reactions, moves priority groups to the defines file

LemonInTheDark committed:
whoops

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into lemon-gets-mad-at-react

LemonInTheDark committed:
Converts air_contents to _air_contents, replaces all the out of file things that used it with return_air()

LemonInTheDark committed:
Replaces the canister air contents uses, converts it back to air_contents, I decided I didn't like the _

LemonInTheDark committed:
Fixes
